### DISCUSSION TITLE
Mutual TLS, different certificate for client/server

### DISCUSSION THREAD

**Discussion by kscieslinski** [mitmproxy/mitmproxy#6264] [Q&A] [answered]
Hi, I have a standard application and a server performing Mutual TLS.

Hopefully I am able to:
i) retrieve both the client certificate and private key,
ii) set a trusted authority for the client,

I was thinking of a setup where I add mitmproxy-ca-cert.pem to the client trusted authorities.
Now I would like mitmproxy to:
- communicate with the client using the autogenerated mitmproxy certificate with mitmproxy-ca-cert.pem,
- communicate with the server using the retrieved client certificate.

Unfortunately I must be missing some understanding of mitmproxy options, as when I use:
```
./mitmdump --ssl-insecure --certs server_hostname=./retrieved-client-certs/client_cert_key.pem --set client_certs=~/.mitmproxy/
```
the mitmproxy keeps communicating with the client using the same certificates as used for communication with the server.

**Comment by mhils** [comment]
What you are doing looks about right! Have you tried to hardcode one specific certificate for `client <-> mitmproxy` communication by specifying `--certs ./retrieved-client-certs/client_cert_key.pem`? For the upstream portion, you shoulsd definitely be using `--set` and not `-set`. You can try a fixed file there as well. That would exclude at least some possible issues why it's not working yet.

**Comment by kscieslinski** [comment, marked as answer]
Thanks for such a quick answer. It turned out that I was wrongly understanding which side is the `client` and which side is the `server` in the mitmproxy nomenclature. Simply switching the arguments around worked :)
